docs/guides/virtualization/cloud-init/03_configuration_engine.md

```diff
@@ -0,0 +1,171 @@
+
+### **Module Deep Dive: `packages`**
+
+The `packages` module provides a declarative way to manage the software on your instance, ensuring that specific applications are installed at boot time.
+
+In this example, we will ensure two useful tools, `nginx` (a high-performance web server) and `htop` (an interactive process viewer), are installed. We will also instruct `cloud-init` to update the package repository metadata first to ensure it can find the latest versions.
+
+1.  **Create `user-data.yml`:**
+    ```bash
+    cat <<EOF > user-data.yml
+    #cloud-config
+    package_update: true
+    packages:
+      - nginx
+      - htop
+    EOF
+    ```
+
+2.  **Key Directives Explained:**
+    *   `package_update: true`: Instructs the package manager to refresh its local metadata. On Rocky Linux, this is equivalent to running `dnf check-update`.
+    *   `packages`: A list of package names to be installed.
+
+3.  **Boot and Verify:** After booting, SSH in and check that `nginx` is installed with `rpm -q nginx`.
+
+!!! note "Idempotency in Action"
+    If you were to reboot this VM with the same `user-data`, the `packages` module would see that `nginx` and `htop` are already installed and do nothing further. It ensures the desired state (packages are present) without taking unnecessary action. That is idempotency.
+
+### **Module Deep Dive: `write_files`**
+
+This module is incredibly versatile, allowing you to write any text content to any file on the system. It is the perfect tool for deploying application configuration files, populating web content, or creating helper scripts.
+
+To demonstrate its power, we will use `write_files` to create a custom homepage for the `nginx` web server that we are also installing in the same run.
+
+1.  **Create `user-data.yml`:**
+    ```bash
+    cat <<EOF > user-data.yml
+    #cloud-config
+    packages: [nginx]
+    write_files:
+      - path: /usr/share/nginx/html/index.html
+        content: '<h1>Hello from cloud-init!</h1>'
+        owner: nginx:nginx
+        permissions: '0644'
+    runcmd:
+      - [ systemctl, enable, --now, nginx ]
+    EOF
+    ```
+
+2.  **Key Directives Explained:**
+    *   `path`: The absolute path on the filesystem where the file will be written.
+    *   `content`: The text content to be written to the file.
+    *   `owner`: Specifies the user and group that should own the file (e.g., `nginx:nginx`).
+    *   `permissions`: The file permissions in octal format (e.g., `0644`).
+
+3.  **Boot and Verify:** After booting, SSH in and use `curl localhost` to see the new homepage.
+
+!!! tip "Writing Binary Files"
+    The `write_files` module is not limited to text. By specifying an `encoding`, you can deploy binary files. For example, you can use `encoding: b64` to write base64-encoded data. For advanced use cases, refer to the [official `write_files` documentation](https://cloudinit.readthedocs.io/en/latest/topics/modules.html#write-files).
```

Comment:
Once again, the admonition needs an extra LF after the title line.  It looks like you need that on all of your admonitions in each of the documents here.